pkg/core/blockchain.go

```diff
@@ -349,9 +360,11 @@ func (bc *Blockchain) AddHeaders(headers ...*block.Header) (err error) {
 			if int(h.Index) < headerList.Len() {
 				continue
 			}
-			if !h.Verify() {
-				err = fmt.Errorf("header %v is invalid", h)
-				return
+			if verify {
+				if err = bc.verifyHeader(h, lastHeader); err != nil {
```

Comment:
I don't think it's a good idea doing verification inside `headersOp` as it effectively locks `bc.headerList` and verification can take some substantial amount of time.

Reply:
I have moved verification of external input upwards, and inside locked context we can only check for the validity of the first header.